components/style/viewport.rs

```diff
@@ -32,6 +35,65 @@ pub enum ViewportDescriptor {
     Orientation(Orientation)
 }
 
+// ViewportLength is a length | percentage | auto | extend-to-zoom
+// See:
+// * http://dev.w3.org/csswg/css-device-adapt/#min-max-width-desc
+// * http://dev.w3.org/csswg/css-device-adapt/#extend-to-zoom
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ViewportLength {
+    Length(Length),
+    Percentage(f32),
+    Auto,
+    ExtendToZoom
+}
+
+impl ToCss for ViewportLength {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result
+        where W: fmt::Write
+    {
+        match self {
+            &ViewportLength::Length(length) => length.to_css(dest),
+            &ViewportLength::Percentage(percentage) => write!(dest, "{}%", percentage * 100.),
+            &ViewportLength::Auto => write!(dest, "auto"),
+            &ViewportLength::ExtendToZoom => write!(dest, "extend-to-zoom"),
+        }
+    }
+}
+
+impl ViewportLength {
+    fn parse(input: &mut Parser) -> Result<ViewportLength, ()> {
+        match try!(input.next()) {
+            Token::Dimension(ref value, ref unit) if AllowedNumericType::NonNegative.is_ok(value.value) =>
+                Length::parse_dimension(value.value, unit).map(ViewportLength::Length),
+            Token::Percentage(ref value) if AllowedNumericType::NonNegative.is_ok(value.unit_value) =>
+                Ok(ViewportLength::Percentage(value.unit_value)),
+            Token::Number(ref value) if value.value == 0. =>
+                Ok(ViewportLength::Length(Length::Absolute(Au(0)))),
+            Token::Ident(ref value) if value.eq_ignore_ascii_case("auto") =>
+                Ok(ViewportLength::Auto),
+            // we explicitly do not accept 'extend-to-zoom', since it is a UA internal value
+            // for <META> viewport translation
+            _ => Err(())
+        }
+    }
+
+    fn from_meta<'a>(value: &'a str) -> Option<ViewportLength> {
+        Some(match value {
+            v if v.eq_ignore_ascii_case("device-width") =>
+                ViewportLength::Length(Length::ViewportPercentage(ViewportPercentageLength::Vw(100.))),
+            v if v.eq_ignore_ascii_case("device-height") =>
+                ViewportLength::Length(Length::ViewportPercentage(ViewportPercentageLength::Vh(100.))),
+            _ => {
+                match value.parse::<f32>() {
+                    Ok(n) if n >= 0. => ViewportLength::Length(Length::from_px(n.max(1.).min(10000.))),
+                    Ok(n) if n < 0. => return None,
+                    _ => ViewportLength::Length(Length::from_px(1.))
```

Comment:
nit: Might be clearer as `Ok(n) if n >= 0. => ..., Ok(n) => ..., Err(_) => ...` ?